rosetta/services/transaction_construction.go

```diff
@@ -0,0 +1,143 @@
+package services
+
+import (
+	"fmt"
+
+	"github.com/coinbase/rosetta-sdk-go/types"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/pkg/errors"
+
+	hmyTypes "github.com/harmony-one/harmony/core/types"
+	"github.com/harmony-one/harmony/rosetta/common"
+)
+
+// ConstructTransaction object (unsigned).
+// TODO (dm): implement staking transaction construction
+func ConstructTransaction(
+	components *OperationComponents, metadata *ConstructMetadata, sourceShardID uint32,
+) (response hmyTypes.PoolTransaction, rosettaError *types.Error) {
+	if components == nil || metadata == nil {
+		return nil, common.NewError(common.CatchAllError, map[string]interface{}{
+			"message": "nil components or metadata",
+		})
+	}
+	if metadata.Transaction.FromShardID != nil && *metadata.Transaction.FromShardID != sourceShardID {
+		return nil, common.NewError(common.InvalidTransactionConstructionError, map[string]interface{}{
+			"message": fmt.Sprintf("intended source shard %v != tx metadata source shard %v",
+				sourceShardID, *metadata.Transaction.FromShardID),
+		})
+	}
+
+	var tx hmyTypes.PoolTransaction
+	switch components.Type {
+	case common.CrossShardTransferOperation:
```

Comment:
why does crossshardtransfer different from plaintransfer?

Reply:
The assumed metadata provided is different. Cross shard has to provide from & to shard in the metadata whereas plain transactions do not. More rosetta implementation do not require metadata for transfers, thus the decision to do it this way.